Modules/System/AL Rest Client/src/ALRestClientImpl.Codeunit.al

```diff
@@ -0,0 +1,153 @@
+        Initialized: Boolean;
+        NotInitializedErr: Label 'The AL Rest Client has not been initialized';
+        EnvironmentBlocksErr: label 'Environment blocks an outgoing HTTP request to ''%1''.', Comment = '%1 = url, e.g. https://microsoft.com';
+        ConnectionErr: label 'Connection to the remote service ''%1'' could not be established.', Comment = '%1 = url, e.g. https://microsoft.com';
+        RequestFailedErr: label 'The request failed: %1 %2', Comment = '%1 = HTTP status code, %2 = Reason phrase';
+        UserAgentLbl: Label 'Dynamics 365 Business Central - |%1| %2/%3', Locked = true, Comment = '%1 = App Publisher; %2 = App Name; %3 = App Version';
+
+    procedure Initialize()
+    begin
+        ClearAll();
+
+        HttpClient.Clear();
+        HttpClient.Timeout := 60000;
+        Initialized := true;
+        SetDefaultUserAgentHeader();
+    end;
+
+    procedure SetBaseAddress(Url: Text)
+    begin
+        AssertInitialized();
+        HttpClient.SetBaseAddress(Url);
+    end;
+
+    procedure GetBaseAddress() Url: Text
+    begin
+        AssertInitialized();
+        Url := HttpClient.GetBaseAddress;
+    end;
+
+    procedure SetTimeOut(TimeOut: Duration)
+    begin
+        AssertInitialized();
+        HttpClient.Timeout := TimeOut;
+    end;
+
+    [NonDebuggable]
+    procedure SetAuthorizationHeader(Value: Text)
+    begin
+        SetDefaultRequestHeader('Authorization', Value);
+    end;
+
+    procedure SetUserAgentHeader(Value: Text)
+    begin
+        SetDefaultRequestHeader('User-Agent', Value);
+    end;
+
+    local procedure SetDefaultUserAgentHeader()
+    begin
+        SetUserAgentHeader(GetUserAgentString());
+    end;
+
+    local procedure GetUserAgentString() UserAgentString: Text
+    var
+        ModuleInfo: ModuleInfo;
+    begin
+        if NavApp.GetCurrentModuleInfo(ModuleInfo) then
+            UserAgentString := StrSubstNo(UserAgentLbl, ModuleInfo.Publisher(), ModuleInfo.Name(), ModuleInfo.AppVersion());
+    end;
+
+    procedure SetDefaultRequestHeader(Name: Text; Value: Text)
+    begin
+        AssertInitialized();
+        if HttpClient.DefaultRequestHeaders.Contains(Name) then
+            HttpClient.DefaultRequestHeaders.Remove(Name);
+        HttpClient.DefaultRequestHeaders.Add(Name, Value);
+    end;
+
+    [NonDebuggable]
+    procedure SetAuthentication(HttpAuthenticationInstance: Interface "Http Authentication")
+    begin
+        HttpAuthentication := HttpAuthenticationInstance;
+    end;
+
+    procedure HasConnectionError(): Boolean
+    begin
+        exit(ConnectionError);
+    end;
+
+    local procedure AssertInitialized()
+    begin
+        if not Initialized then
+            Error(NotInitializedErr);
+    end;
+
+    procedure SendRequest(var ALHttpRequestMessage: Codeunit "AL Http Request Message") ALHttpResponseMessage: Codeunit "AL Http Response Message"
+    begin
+        if not TrySendRequest(ALHttpRequestMessage, ALHttpResponseMessage) then;
```

Comment:
This is up for discussion. I chose to catch the errors here and handle them later. The procedure Send will throw the error when there is a connection problem. But I believe that http status codes should not be handled in any way, that's up to the caller.
The reason for encapsulating it in a TryFunction is to throw the correct error message here, so calling functions can grab GetLastErrorText. At the same time, the error message is also stored in the "AL Http Response Message" object, so that is a duplicate.

Reply:
After some consideration, I've decided to remove the TryFunction and not throw an error at all. Instead, the SendRequest function returns true or false and the error message is inside the response codeunit.
An error occurring during setting the authorization header now throws an error immediately. That's considered to be a runtime error, something I can't handle.
If the outgoing call is blocked by the environment or fails because of no connection to the server, then this will also result in an error.
As a result, only when a response was received there will be no error message. This is similar to how the HttpClient works out of the box.